_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+
+With this data, it's easy to imagine a bunch of interesting data-analysis questions; however here our focus is on differential privacy, so we'll keep the setup simple. We will try to predict heart rate given the physiological measurements from one of the two devices, Empatica E4. Also, we'll zoom in on a single subject, _S1_, who will provide us with 4603 instances of two-second heart rate values. ^[Relative files per subject are 1.4G in size.]
+
+As usual, we start with the required libraries; unusually though, as of this writing we need to disable version 2 behavior in TensorFlow, as TensorFlow Privacy does not yet fully work with TF 2. (Hopefully, for many future readers, this won't be the case anymore.)
+Note how TF Privacy -- a Python library -- is imported via `reticulate`.
+
+```{r}
+library(tensorflow)
+tf$compat$v1$disable_v2_behavior()
+
+library(keras)
+library(tfdatasets)
+library(tfautograph)
+
+library(purrr)
+
+library(reticulate)
+# if you haven't yet, install TF Privacy, e.g. using reticulate:
+# py_install("tensorflow_privacy")
+priv <- import("tensorflow_privacy")
+```
+
+From the downloaded archive, we just need `S1.pkl`, saved in a [native Python serialization format](https://docs.python.org/3.8/library/pickle.html), yet nicely loadable using `reticulate`:
+
+```{r}
+s1 <- py_load_object("PPG_FieldStudy/S1/S1.pkl", encoding = "latin1")
+```
+
+`s1` points to an R list comprising elements of different length -- the various physical/physiological signals have been sampled with different frequencies:
+
+```{r}
+### predictors ###
+
+# accelerometer data - sampling freq. 32 Hz
+# also note that these are 3 "columns", for each of x, y, and z axes
+s1$signal$wrist$ACC %>% nrow() # 294784
+# PPG data - sampling freq. 64 Hz
+s1$signal$wrist$BVP %>% nrow() # 589568
+# electrodermal activity data - sampling freq. 4 Hz
+s1$signal$wrist$EDA %>% nrow() # 36848
+# body temperature data - sampling freq. 4 Hz
+s1$signal$wrist$TEMP %>% nrow() # 36848
+
+### target ###
+
+# EKG data - provided in already averaged form, at frequency 0.5 Hz
+s1$label %>% nrow() # 4603
+```
+
+In light of the different sampling frequencies, our `tfdatasets` pipeline will have do some moving averaging, paralleling that applied to construct the ground truth data.
+
+
+## Preprocessing pipeline
+
+As every "column" ^[For convenience, we take the liberty to talk as if we had the usual rectangular data here.] is of different length and resolution, we build up the final _dataset_ piece-by-piece.
+The following function serves two purposes:
+
+1. compute running averages over differently sized windows, thus downsampling to 0.5Hz for every modality
+2. transform the data to the `(num_timesteps, num_features)` format that will be required by the 1d-convnet we're going to use soon
+
+```{r}
+average_and_make_sequences <-
+  function(data, window_size_avg, num_timesteps) {
+    data %>% tf$cast(tf$float32) %>%
```

Comment:
Somewhat off-topic, do you think it's acceptable style to use `k_` functions where available to avoid `$`? E.g. `k_cast()` instead of `tf$cast()`.

Reply:
yeah you may be right, it probably is preferred style-wise (?)
I think I got "spoiled" by the TFP applications not involving `keras`, so I tend to automatically use `tf$` style sometimes...
Changing.